sound/soc/sof/control.c

```diff
@@ -341,25 +345,53 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 	struct snd_sof_control *scontrol = be->dobj.private;
 	struct snd_sof_dev *sdev = scontrol->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	unsigned int tag = 0;
+	struct snd_ctl_tlv header;
+	struct snd_ctl_tlv *tlvd = (struct snd_ctl_tlv *)binary_data;
+	int max_length = SOF_IPC_MSG_MAX_SIZE -
+		sizeof(const struct sof_ipc_ctrl_data) -
+		sizeof(const struct sof_abi_hdr);
+	int ret;
 
-	pm_runtime_get_sync(sdev->dev);
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: bytes get failed to resume %d\n",
+			ret);
+		return ret;
+	}
 
-	dev_dbg(sdev->dev, "getting data and command is %d\n", scontrol->cmd);
-	/* get all the mixer data from DSP */
-	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_DATA,
-				  SOF_CTRL_TYPE_DATA_GET, scontrol->cmd);
+	/* Decrement size to fit the ext bytes header and get the the
+	 * upper limit from ext bytes control size from topology and
+	 * SOF IPC max. size limit.
+	 */
+	size -= sizeof(const struct snd_ctl_tlv);
+	if (max_length > be->max)
+		max_length = be->max;
 
-	/* TODO: replace 252 with actual size */
-	if (copy_to_user(binary_data, &tag, sizeof(u32)))
-		return -EFAULT;
-	if (copy_to_user(binary_data + 1, &size, sizeof(u32)))
+	dev_dbg(sdev->dev, "request size minus header is %d\n", size);
+	dev_dbg(sdev->dev, "max size is %d\n", max_length);
+
+	/* Prevent read of other kernel data */
+	if (size > max_length) {
+		dev_err(sdev->dev, "error: size is too large\n");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* get all the component data from DSP */
+	ret = snd_sof_ipc_get_comp_data(sdev->ipc, scontrol,
+					SOF_IPC_COMP_GET_DATA,
+					SOF_CTRL_TYPE_DATA_GET, scontrol->cmd);
+
+	header.numid = 0;
```

Comment:
Should't this be the same ID as passed in binary_data ?

Reply:
Do you mean that numid should be set to scontrol->cmd (e.g. SOF_CTRL_CMD_BINARY)? And should we add check to put function that numid == scontrol->cmd?